File: Dynamic-Programming-I/P392.cpp
// Date: Tue, Sep 21, 2021
// 392. Is Subsequence (https://leetcode.com/problems/is-subsequence/)

// Example 1:
// Input: s = "abc", t = "ahbgdc"
// Output: true

// Example 2:
// Input: s = "axc", t = "ahbgdc"
// Output: false

bool isSubsequence(string s, string t) {
	int n_t = t.length();
	int n_s = s.length();
	int p_s = 0;
	int p_t = 0;
	while(p_t < n_t && p_s < n_s) {
		if (s[p_s] == t[p_t])
			p_s++;
		p_t++;
	}
	return p_s == n_s;
}

File: LeetCoding-Challenge/March-2022/P392.java
package leetcodingchallenge.march2022;

/**
 * Date: March 02, 2022
 * 392. Is Subsequence
 * https://leetcode.com/problems/is-subsequence/
 */

public class P392 {
    // time: O(n)
    // space: O(1)
    public boolean isSubsequence(String s, String t) {
        int n1 = s.length(), n2 = t.length();
        if (n1 > n2) return false;
        int p1 = 0, p2 = 0;
        while (p1 < n1 && p2 < n2) p1 += s.charAt(p1) == t.charAt(p2++) ? 1 : 0;
        return p1 == n1;
    }
}
